lib/transactions/sendAndConfirm.ts

```diff
@@ -0,0 +1,136 @@
+ *
+ * @returns {Promise<string>} - A promise that resolves with the transaction ID if the transaction is confirmed successfully, or rejects with an error if the transaction fails or timeouts.
+ *
+ * @throws {ConfirmationTimeoutError} - Throws this error if the confirmation process times out.
+ * @throws {InstructionError} - Throws this error if there is an issue with the transaction's instructions.
+ * @throws {TransactionError} - Throws this error for general transaction failures.
+ */
+
+export const sendAndConfirmTransaction = async ({
+  signedTransaction,
+  connection,
+  config = {
+    type: 'static',
+    timeoutMs: DEFAULT_CONFIRMATION_TIMEOUT,
+  },
+  onTransactionEvent,
+}: {
+  signedTransaction: Transaction | VersionedTransaction;
+  connection: Connection;
+  config?: SendTransactionConfig &
+    (
+      | StaticTimeoutConfig
+      | TransactionExpirationTimeoutConfig
+      | NoTimeoutConfig
+    );
+  onTransactionEvent?: TransactionLifecycleEventCallback;
+}): Promise<string> => {
+  const startTime = getUnixTs();
+  const controller = new AbortController();
+
+  const {
+    continuouslySendTransactions,
+    pollingSendTransactionTimeoutMs,
+    sendOptions,
+    ...confirmationConfig
+  } = config;
+
+  const transactionId = await sendSignedTransaction({
+    signedTransaction,
+    connection,
+    config: {
+      controller,
+      continuouslySendTransactions,
+      pollingSendTransactionTimeoutMs,
+      sendOptions,
+    },
+    onTransactionEvent,
+  });
+
+  if (confirmationConfig.type === 'none') {
+    controller.abort();
+    log(
+      'Caller requested no confirmation, skipping all confirmation and returning after initial transaction sent to cluster'
+    );
+
+    return transactionId;
+  }
+
+  try {
+    await awaitTransactionSignatureConfirmation({
+      connection,
+      transactionId,
+      config: confirmationConfig,
+      onTransactionEvent,
+      controller,
+      transactionCommitment:
+        confirmationConfig?.type === 'expiration'
+          ? confirmationConfig?.transactionCommitment
+          : undefined,
+    });
+
+    log(
+      'Finished transaction status confirmation: ',
+      transactionId,
+      getUnixTs() - startTime
+    );
+  } catch (error: any) {
+    if (error.timeout) {
+      throw new ConfirmationTimeoutError({
+        transactionId,
+        message: 'Timed out awaiting confirmation on transaction',
+        config: ConfirmationTimeoutError.formatConfig(confirmationConfig),
+      });
+    } else if (error.err) {
+      throw new InstructionError({
+        transactionId,
+        error: error.err,
+        transaction: signedTransaction,
+      });
+    }
+
+    // note: if we want to try and get more information about a transaction failure, we could make a simulate transaction request here
+
+    // question: is there any additional processing we can do to give back more information to the caller?
```

Comment:
Indeed, depends on that triggers this. However if we're subscribed we may be able to subscribe logs for this txn and parse the logs, that's the only way, unless there's something I've missed before.

Also FYI, not all RPCs process log messages, so that's a fun nuance ;)

Reply:
I'm not sure about the subscription case — I see there's an [onLogs](https://solana-labs.github.io/solana-web3.js/classes/Connection.html#onLogs) web socket, but it looks like it's for a specific public key? BUT, if we only care about trying to parse logs on failed transactions, we could get those (kinda for free) if we want to simulate the failed transaction anyways 🙂 

I didn't know they _might_ be there depending on the RPC, so thanks for pointing out! 🙌